### PR TITLE
Tracing

### PR DESCRIPTION
**Pull request by agneborn98**
Basic tracing implementation.

For future logging, use the _level_! macro to send an event to the tracer, as outlined in [Shorthand Macros](https://docs.rs/tracing/0.1.35/tracing/#shorthand-macros).

### CODE REVIEW

**Comment by CGodiksen** [comment]
You mentioned the "event!" macro but linked to some alternatives to that macro. Should we use "event!" with a manual log level or use the macros with the predefined log level?

**Comment by agneborn98** [comment]
Yeah okay... I can see why that explanation is confusing. What I meant is that you should use the _level_! macro to make events. Maybe just refer to the documentation, it is the better explanation.

File: server/src/catalog.rs [hunk 1]
```diff
@@ -138,7 +142,7 @@ fn is_dir_entry_a_table(dir_entry: &DirEntry) -> bool {
 }
 
 fn is_dir_entry_a_parquet_file(dir_entry: &DirEntry) -> bool {
-    let mut file = File::open(dir_entry.path()).unwrap();
+    let mut file = OpenOptions::new().write(true).open(dir_entry.path()).unwrap();
```
Comment (CGodiksen) [inline server/src/catalog.rs:145]: Is this change relevant to tracing? If not it should be reverted and handled in a separate branch. As far as i remember this is the fix to the Windows issue so it should probably be in an individual bug fix PR.

**Comment by agneborn98** [comment]
To further answer your question, it doesn't really matter if you use the manual log level with event! or use the level! macro, they are the same. My suggestion was just for consistency's sake.

File: server/src/catalog.rs [hunk 2]
```diff
@@ -17,6 +17,9 @@ use std::io::Read;
 use std::sync::Arc;
 use std::{fs::File, path::Path};
 use std::str;
+use std::fs::OpenOptions;
+
+use tracing::{debug, error, info, warn, Level, event, instrument, span};
```
Comment (CGodiksen) [inline server/src/catalog.rs:22]: This will potentially be handled automatically by rustfmt in the future but for now we group imports manually as std -> external crates -> internal crates. The tracing import should be moved to the same group as datafusion (which should be one group).

File: server/src/main.rs [hunk 3]
```diff
@@ -21,7 +21,12 @@ mod tables;
 mod types;
 mod storage;
 
-use std::sync::Arc;
```
Comment (CGodiksen) [inline server/src/main.rs:24]: There should only be one newline between modules and imports.

File: server/src/main.rs [hunk 4]
```diff
@@ -21,7 +21,12 @@ mod tables;
...
+
+use std::{fs::File, sync::Arc, fmt};
```
Comment (CGodiksen) [inline server/src/main.rs:25]: As mentioned in the comment to `catalog.rs`  we should group import manually as std -> external crates -> internal crates. tracing, tokio and datafusion should therefore be in one group.

File: server/src/main.rs [hunk 5]
```diff
@@ -33,6 +38,7 @@ use crate::catalog::Catalog;
 use crate::optimizer::model_simple_aggregates;
 use crate::tables::ModelTable;
 
+
```
Comment (CGodiksen) [inline server/src/main.rs:41]: There should only be one newline between imports and the rest of the code.

File: server/src/main.rs [hunk 6]
```diff
@@ -45,13 +51,38 @@ pub struct Context {
 
 /** Public Functions **/
 fn main() {
+    
+
```
Comment (CGodiksen) [inline server/src/main.rs:55]: There should not be any newlines in the start of a function.

File: server/src/main.rs [hunk 7]
```diff
@@ -45,13 +51,38 @@ pub struct Context {
...
+    // A layer that logs events to a file.
+    let file = File::create("debug.log");
+    let file = match file  {Ok(file) => file,Err(error) => panic!("Error: {:?}",error),};
```
Comment (CGodiksen) [inline server/src/main.rs:58]: This does not seem formatted. Consider running `rustfmt` to fix issues like this.

Comment (CGodiksen) [reply server/src/main.rs:58]: You could also consider just matching on the result of `File::create("debug.log")` instead of using an intermediate variable.

File: server/src/main.rs [hunk 8]
```diff
@@ -45,13 +51,38 @@ pub struct Context {
...
+    let debug_log = tracing_subscriber::fmt::layer()
+        .with_writer(Arc::new(file));
+    
+    // A layer that logs events to stdout.
+
```
Comment (CGodiksen) [inline server/src/main.rs:63]: There should not be a newline between comments and what they relate to.

File: server/src/main.rs [hunk 9]
```diff
@@ -45,13 +51,38 @@ pub struct Context {
...
+    let stdout_log = tracing_subscriber::fmt::layer()
+    .pretty();
```
Comment (CGodiksen) [inline server/src/main.rs:65]: This does also not seem formatted. As mentioned above `rustfmt` should be used.

File: server/src/main.rs [hunk 10]
```diff
@@ -45,13 +51,38 @@ pub struct Context {
...
+    
+    tracing_subscriber::registry()
```
Comment (CGodiksen) [inline server/src/main.rs:67]: Formatting issue.

File: server/src/main.rs [hunk 11]
```diff
@@ -45,13 +51,38 @@ pub struct Context {
...
+        .init();
+    
     let mut args = std::env::args();
     args.next(); // Skip executable.
     if let Some(data_folder) = args.next() {
         // Build Context.
         let catalog = catalog::new(&data_folder);
         let runtime = Runtime::new().unwrap();
         let mut session = create_session_context();
+
```
Comment (CGodiksen) [inline server/src/main.rs:85]: There should not be two newlines here.

File: server/src/main.rs [hunk 12]
```diff
@@ -67,7 +98,7 @@ fn main() {
         // The errors are consciously ignored as the program is terminating.
         let binary_path = std::env::current_exe().unwrap();
         let binary_name = binary_path.file_name().unwrap();
-        println!("usage: {} data_folder", binary_name.to_str().unwrap());
+        error!(" Usage: {} data_folder ", binary_name.to_str().unwrap());
```
Comment (CGodiksen) [inline server/src/main.rs:101]: Why is there an added space at the end of this log message?

File: server/src/main.rs [hunk 13]
```diff
@@ -86,20 +117,21 @@ async fn register_tables(session: &mut SessionContext, catalog: &Catalog) {
     for table_metadata in &catalog.table_metadata {
         if session
             .register_parquet(&table_metadata.name, &table_metadata.path, ParquetReadOptions::default())
+            .instrument(tracing::error_span!("register_parquet"))
             .await
             .is_err()
         {
-            eprintln!("ERROR: unable to initialize table {}", table_metadata.name);
+            error!("ERROR: unable to initialize table {}", table_metadata.name);
```
Comment (CGodiksen) [inline server/src/main.rs:124]: Also consider removing manual log level specifiers here.

File: server/src/main.rs [hunk 14]
```diff
@@ -86,20 +117,21 @@ async fn register_tables(session: &mut SessionContext, catalog: &Catalog) {
...
         }
     }
 
     // Initializes tables storing time series as models in Apache Parquet files.
     for model_table_metadata in &catalog.model_table_metadata {
-        if session
+            if session
```
Comment (CGodiksen) [inline server/src/main.rs:130]: Formatting issue.

File: server/src/remote.rs [hunk 15]
```diff
@@ -22,6 +22,8 @@ use crate::Context;
 
 use std::convert::TryInto;
 
+use tracing::{debug, error, info, warn, Level, event, instrument, span};
```
Comment (CGodiksen) [inline server/src/remote.rs:25]: Imports should be grouped as in `catalog.rs` and `main.rs`.

File: server/src/remote.rs [hunk 16]
```diff
@@ -108,7 +110,6 @@ impl FlightService for FlightServiceHandler {
     ) -> Result<Response<FlightInfo>, Status> {
         Err(Status::unimplemented("Not yet implemented"))
     }
-
```
Comment (CGodiksen) [inline server/src/remote.rs:111]: There should be a newline between functions.

File: server/src/tables.rs [hunk 17]
```diff
@@ -21,6 +21,8 @@ use std::task::{Context, Poll};
 
 use async_trait::async_trait;
 
+use tracing::{debug, error, info, warn, Level, event, instrument, span};
```
Comment (CGodiksen) [inline server/src/tables.rs:24]: Imports should be grouped as in `catalog.rs`, `main.rs`, and `remote.rs`.

File: server/src/catalog.rs [hunk 18]
```diff
@@ -74,7 +79,6 @@ pub fn new(data_folder: &str) -> Catalog {
         Field::new("model", DataType::Binary, false),
         Field::new("gaps", DataType::Binary, false),
     ]));
-
```
Comment (CGodiksen) [inline server/src/catalog.rs:77]: There should probably be a newline here.

Comment (agneborn98) [reply server/src/catalog.rs:77]: Fixed

File: server/src/main.rs (hunk 11 above)
Comment (agneborn98) [reply server/src/main.rs:85]: Fixed

File: server/src/main.rs (hunk 12 above)
Comment (agneborn98) [reply server/src/main.rs:101]: Fixed

File: server/src/remote.rs (hunk 16 above)
Comment (agneborn98) [reply server/src/remote.rs:111]: Fixed

File: server/src/main.rs (hunk 8 above)
Comment (agneborn98) [reply server/src/main.rs:63]: Fixed

File: server/src/main.rs (hunk 6 above)
Comment (agneborn98) [reply server/src/main.rs:55]: Fixed

File: server/src/main.rs (hunk 5 above)
Comment (agneborn98) [reply server/src/main.rs:41]: Fixed

File: server/src/main.rs (hunk 3 above)
Comment (agneborn98) [reply server/src/main.rs:24]: Fixed

File: server/Cargo.toml [hunk 19]
```diff
@@ -10,7 +10,13 @@ datafusion = "10.0.0"
 datafusion-physical-expr = "10.0.0"
 object_store = "0.3.0"
 
+
```
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/Cargo.toml:13]: Fixed

File: server/src/catalog.rs (hunk 1 above)
Comment (agneborn98) [reply server/src/catalog.rs:145]: Fixed

File: server/src/main.rs (hunk 10 above)
Comment (agneborn98) [reply server/src/main.rs:62]: Fixed, rustfmt was run

File: server/src/catalog.rs (hunk 2 above)
Comment (agneborn98) [reply server/src/catalog.rs]: It looks like std -> external crates -> internal crates to me?

Comment (agneborn98) [reply server/src/catalog.rs]: Fixed

File: server/src/main.rs (hunk 4 above)
Comment (agneborn98) [reply server/src/main.rs]: Fixed

File: server/src/main.rs (hunk 9 above)
Comment (agneborn98) [reply server/src/main.rs]: Fixed, rustfmt has run

File: server/src/main.rs (hunk 14 above)
Comment (agneborn98) [reply server/src/main.rs]: Fixed, rustfmt has run

File: server/src/tables.rs (hunk 17 above)
Comment (agneborn98) [reply server/src/tables.rs]: Fixed

File: server/src/remote.rs (hunk 15 above)
Comment (agneborn98) [reply server/src/remote.rs]: Fixed

File: server/src/main.rs (hunk 7 above)
Comment (agneborn98) [reply server/src/main.rs]: Formatted, rustfmt has run

File: server/src/catalog.rs [hunk 20]
```diff
@@ -85,31 +89,31 @@ pub fn new(data_folder: &str) -> Catalog {
...
                             normalized_file_name,
                             path.to_string(),
                             &model_table_segment_group_file_schema,
                         ) {
                             model_table_metadata.push(mtd);
-                            eprintln!("INFO: initialized model table {}", path);
+                            info!("INFO: initialized model table {}", path);
                         } else {
-                            eprintln!("ERROR: unsupported model table {}", path);
+                            error!("ERROR: unsupported model table {}", path);
```
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/catalog.rs:106]: Fixed

File: server/src/catalog.rs [hunk 21]
```diff
@@ -85,31 +89,31 @@ pub fn new(data_folder: &str) -> Catalog {
                     let normalized_file_name = file_name.to_ascii_lowercase();
                     if is_dir_entry_a_table(&dir_entry) {
                         table_metadata.push(new_table_metadata(normalized_file_name, path.to_string()));
-                        eprintln!("INFO: initialized table {}", path);
+                        info!("INFO: initialized table {}", path);
```
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/catalog.rs:96]: Fixed

File: server/src/main.rs (hunk 7 above)
Comment (agneborn98) [reply server/src/main.rs]: 2nd comment is also fixed.

File: server/src/catalog.rs (hunk 1 above)
Comment (CGodiksen) [reply server/src/catalog.rs]: @skejserjensen Should we fix this bug now or will it be handled later as part of another change?

File: server/Cargo.toml (hunk 19 above)
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/Cargo.toml]: @skejserjensen Should we group dependencies and if so, how should they be grouped?

File: server/src/remote.rs (hunk 15 above)
Comment (CGodiksen) [reply server/src/remote.rs]: I am not sure this is fixed.

File: server/src/tables.rs (hunk 17 above)
Comment (CGodiksen) [reply server/src/tables.rs]: I am not sure this is fixed.

File: server/src/catalog.rs [hunk 22]
```diff
@@ -88,7 +89,8 @@ pub fn new(data_folder: &str) -> Catalog {
                     // HACK: workaround for datafusion 8.0.0 lowercasing table names in queries.
                     let normalized_file_name = file_name.to_ascii_lowercase();
                     if is_dir_entry_a_table(&dir_entry) {
-                        table_metadata.push(new_table_metadata(normalized_file_name, path.to_string()));
+                        table_metadata
+                            .push(new_table_metadata(normalized_file_name, path.to_string()));
                         info!("INFO: initialized table {}", path);
```
Comment (CGodiksen) [inline server/src/catalog.rs:94]: Manual log level specifier should be removed here as well.

File: server/src/catalog.rs [hunk 23]
```diff
@@ -97,23 +99,23 @@ pub fn new(data_folder: &str) -> Catalog {
                             &model_table_segment_group_file_schema,
                         ) {
                             model_table_metadata.push(mtd);
-                            info!("INFO: initialized model table {}", path);
+                            info!("initialized model table {}", path);
```
Comment (CGodiksen) [inline server/src/catalog.rs:102]: @skejserjensen We should agree on the convention for writing log messages. I have been using full sentences with punctuation.

File: server/src/remote.rs (hunk 15 above)
Comment (agneborn98) [reply server/src/remote.rs]: You are absolutely right. Now it should be fixed.

File: server/src/tables.rs (hunk 17 above)
Comment (agneborn98) [reply server/src/tables.rs]: You are absolutely right. Now it should be fixed.

**Review by CGodiksen — CHANGES_REQUESTED** [review]

File: server/src/main.rs [hunk 24]
```diff
@@ -67,7 +89,7 @@ fn main() {
         // The errors are consciously ignored as the program is terminating.
         let binary_path = std::env::current_exe().unwrap();
         let binary_name = binary_path.file_name().unwrap();
-        println!("usage: {} data_folder", binary_name.to_str().unwrap());
+        error!(" Usage: {} data_folder", binary_name.to_str().unwrap());
```
Comment (skejserjensen) [inline server/src/main.rs:92]: Should the usage message really be a log message as it simply informs the user what arguments are required to start the system? For example for Apache Cassandra the usage message is:

 `Usage: ./cassandra [-f] [-h] [-p pidfile] [-H dumpfile] [-E errorfile]`

Despite using formatted logging through SLF4J at runtime. I propose we write usage to `eprintln!`.

**Review by skejserjensen — CHANGES_REQUESTED** [review]
I purposely did not comment on the use of `println!` for logging in the `storage` module as I assume this will be changed during development of the module itself.

File: server/src/catalog.rs (hunk 1 above)
Comment (skejserjensen) [reply server/src/catalog.rs]: I think we should fix it in a separate branch now as I do not know when the `catalog` module will be updated. Also we should probably include a comment that documents why a file that is only read is opened with `write(true)`.

File: server/Cargo.toml (hunk 19 above)
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/Cargo.toml]: I simply grouped them based on their functionality because it seemed intuitive to me, e.g., `arrow-flight` and `tonic` together because `arrow-flight` uses `tonic` for gRPC. However, if @agneborn98 and you (@CGodiksen) think another ordering i better, e.g., a simple sorted list, then that is fine by me.

File: server/src/catalog.rs (hunk 1 above)
Comment (CGodiksen) [reply server/src/catalog.rs]: Alright, I can do this once the tracing branch is merged.

File: server/src/main.rs [hunk 25]
```diff
@@ -45,6 +46,27 @@ pub struct Context {
 
 /** Public Functions **/
 fn main() {
+    // A layer that logs events to a file.
+    let file = match File::create("debug.log") {
```
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/main.rs:50]: I agree, there is no reason to add extra overhead to tracing with a filewrite.

File: server/Cargo.toml (hunk 19 above)
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/Cargo.toml]: That grouping seems fine, lets keep doing that.

File: server/src/catalog.rs (hunk 23 above)
Comment (agneborn98) [reply server/src/catalog.rs:102]: Yeah, punctuation looks fine to me. Let's agree on that

**Comment by agneborn98** [comment]
I think all comments have been discussed, I will make the changes ASAP and push it

File: server/src/main.rs (hunk 24 above)
Comment (agneborn98) [reply server/src/main.rs:92]: @CGodiksen Opinion?

File: server/src/catalog.rs (hunk 23 above)
Comment (CGodiksen) [reply server/src/catalog.rs:102]: Great, if you have time you can just change the comments in this branch. I think it still aligns with the overall purpose.

File: server/src/main.rs (hunk 24 above)
Comment (CGodiksen) [reply server/src/main.rs:92]: That sounds fine.

Comment (agneborn98) [reply server/src/main.rs]: Fixed

File: server/src/main.rs [hunk 26]
```diff
@@ -45,6 +46,27 @@ pub struct Context {
...
+        Err(error) => panic!("Error: {:?}", error),
+    };
+    let debug_log = tracing_subscriber::fmt::layer().with_writer(Arc::new(file));
+
+    // A layer that logs events to stdout.
+    let stdout_log = tracing_subscriber::fmt::layer().pretty();
+
+    tracing_subscriber::registry()
+        .with(
+            stdout_log
```
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/main.rs:54]: Fixed

File: server/src/main.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/main.rs]: Fixed

File: server/src/catalog.rs (hunk 23 above)
Comment (agneborn98) [reply server/src/catalog.rs]: Fixed

File: server/src/catalog.rs (hunk 22 above)
Comment (agneborn98) [reply server/src/catalog.rs]: Fixed

File: server/src/catalog.rs [hunk 27]
```diff
@@ -91,31 +91,31 @@ pub fn new(data_folder: &str) -> Catalog {
                     if is_dir_entry_a_table(&dir_entry) {
                         table_metadata
                             .push(new_table_metadata(normalized_file_name, path.to_string()));
-                        info!("INFO: initialized table {}", path);
+                        info!("Initialized table {}", path);
```
Comment (CGodiksen) [inline server/src/catalog.rs:94]: Missing punctuation.

File: server/src/main.rs [hunk 28]
```diff
@@ -89,7 +82,7 @@ fn main() {
         // The errors are consciously ignored as the program is terminating.
         let binary_path = std::env::current_exe().unwrap();
         let binary_name = binary_path.file_name().unwrap();
-        error!(" Usage: {} data_folder", binary_name.to_str().unwrap());
+        println!(" Usage: {} data_folder", binary_name.to_str().unwrap());
```
Comment (CGodiksen) [inline server/src/main.rs:85]: There should not be a space in the beginning of this message. Also should `println!` be used or `eprintln!`?

**Review by CGodiksen — CHANGES_REQUESTED** [review]

File: server/src/main.rs (hunk 28 above)
Comment (CGodiksen) [reply server/src/main.rs:85]: Also as far as i can see, punctuation has only been added in `catalog.rs`. You might as well change all log messages that have been added in this branch.

Comment (agneborn98) [reply server/src/main.rs]: Fixed

File: server/src/catalog.rs (hunk 27 above)
Comment (agneborn98) [reply server/src/catalog.rs]: Fixed

File: server/src/main.rs (hunk 28 above)
Comment (agneborn98) [reply server/src/main.rs]: > There should not be a space in the beginning of this message. Also should `println!` be used or `eprintln!`?

I couldn't tell you the difference between stderr and stdout, I don't know if it is important to the program. Both options print it in the console fine

Comment (CGodiksen) [reply server/src/main.rs]: I think `eprintln!` is correct here.

Comment (agneborn98) [reply server/src/main.rs]: Fixed

**Review by CGodiksen — APPROVED** [review]

File: server/src/main.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/main.rs]: I think we should remove `.pretty()` and `.with_filter(filter::LevelFilter::INFO)` so the format of `stdout_log` matches the single line format the removed `debug_log` wrote to the file as mentioned above. By removing `.with_filter(filter::LevelFilter::INFO)` we can also statically control which log messages are printed completely from the features of the `tracing` crate in `cargo.toml` instead of both in `cargo.toml` and `main.rs`. If we want to have more control over which messages are printed in the future we can always add [`EnvFilter` ](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html) so the filtering level can be set as an environment variable.

File: server/src/main.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/main.rs:54]: As mentioned above I think we should remove `with_filter(filter::LevelFilter::INFO)`.

File: server/src/main.rs (hunk 24 above)
Comment (skejserjensen) [reply server/src/main.rs]: Looks like you accidentally used `println!` instead of `eprintln!`.

Comment (agneborn98) [reply server/src/main.rs]: No, my latest push is eprintln!

File: server/src/main.rs (hunk 28 above)
Comment (skejserjensen) [reply server/src/main.rs]: By keeping normal output in stdout and stderr you can pipe normal output and errors to different places.

File: server/src/main.rs (hunk 24 above)
Comment (skejserjensen) [reply server/src/main.rs]: Sorry my mistake, somehow I looked at a version of the branch without that commit.

File: server/src/main.rs (hunk 26 above)
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/main.rs:54]: Okay, fixed.

File: server/src/main.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (agneborn98) [reply server/src/main.rs]: Done

**Review by CGodiksen — APPROVED** [review]

**Review by skejserjensen — APPROVED** [review]
